internal/core/operations/switch_active_version/switch_active_version_executor.go

```diff
@@ -0,0 +1,197 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package switch_active_version
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/topfreegames/maestro/internal/core/entities"
+	"github.com/topfreegames/maestro/internal/core/entities/game_room"
+	"github.com/topfreegames/maestro/internal/core/entities/operation"
+	"github.com/topfreegames/maestro/internal/core/operations"
+	"github.com/topfreegames/maestro/internal/core/services/interfaces"
+	"github.com/topfreegames/maestro/internal/core/services/room_manager"
+	"go.uber.org/zap"
+)
+
+var roomsSlice []*game_room.GameRoom
+
+type SwitchActiveVersionExecutor struct {
+	roomManager        *room_manager.RoomManager
+	schedulerManager   interfaces.SchedulerManager
+	roomsBeingReplaced *sync.Map
+	newCreatedRooms    []*game_room.GameRoom
+}
+
+func NewExecutor(roomManager *room_manager.RoomManager, schedulerManager interfaces.SchedulerManager) *SwitchActiveVersionExecutor {
+	return &SwitchActiveVersionExecutor{
+		roomManager:        roomManager,
+		schedulerManager:   schedulerManager,
+		roomsBeingReplaced: &sync.Map{},
+		newCreatedRooms:    roomsSlice,
+	}
+}
+
+// Execute the process of switching a scheduler active version consists of the following:
+// 1. Creates "replace" goroutines (same number as MaxSurge);
+// 2. Each goroutine will listen to a channel and create a new room using the
+//    new configuration. After the room is ready, it will then delete the room
+//    being replaced;
+// 3. List all game rooms that need to be replaced and produce them into the
+//    replace goroutines channel;
+// 4. Switch the active version
+func (e *SwitchActiveVersionExecutor) Execute(ctx context.Context, op *operation.Operation, definition operations.Definition) error {
+	logger := zap.L().With(
+		zap.String("scheduler_name", op.SchedulerName),
+		zap.String("operation_definition", definition.Name()),
+		zap.String("operation_id", op.ID),
+	)
+	logger.Debug("start switching scheduler active version")
+
+	updateDefinition := definition.(*SwitchActiveVersionDefinition)
+	scheduler := &updateDefinition.NewActiveScheduler
+
+	maxSurgeNum, err := e.roomManager.SchedulerMaxSurge(ctx, scheduler)
+	if err != nil {
+		return fmt.Errorf("error fetching scheduler max surge: %w", err)
+	}
+
+	roomsChan := make(chan *game_room.GameRoom)
+	var maxSurgeWg sync.WaitGroup
+
+	maxSurgeWg.Add(maxSurgeNum)
+	for i := 0; i < maxSurgeNum; i++ {
+		go e.replaceRoom(logger, &maxSurgeWg, roomsChan, e.roomManager, *scheduler)
+	}
+
+roomsListLoop:
```

Comment:
@caiordjesus What about isolating this loop execution and logic into a separate private method,something like `se.startReplaceLoop(ctx context.Context, scheduler *entities.Scheduler, maxSurgeNum int, roomsChan chan *game_room.GameRoom) error` or `se.executeReplaceLoop(ctx context.Context, scheduler *entities.Scheduler, maxSurgeNum int, roomsChan chan *game_room.GameRoom) error`